clippy_lints/src/new_without_default.rs

```diff
@@ -156,16 +224,128 @@ impl<'tcx> LateLintPass<'tcx> for NewWithoutDefault {
     }
 }
 
-fn create_new_without_default_suggest_msg(
-    self_type_snip: &str,
-    generics_sugg: &str,
-    where_clause_sugg: &str,
-) -> String {
-    #[rustfmt::skip]
-    format!(
-"impl{generics_sugg} Default for {self_type_snip}{where_clause_sugg} {{
+// Check if Self is a unit struct, and avoid any kind of suggestions
+// FIXME: this was copied from DefaultConstructedUnitStructs,
+//        and should be refactored into a common function
+fn is_unit_struct(_cx: &LateContext<'_>, ty: Ty<'_>) -> bool {
+    if let Adt(def, ..) = ty.kind()
+        && def.is_struct()
+        && let var @ VariantDef {
+            ctor: Some((hir::def::CtorKind::Const, _)),
+            ..
+        } = def.non_enum_variant()
+        && !var.is_field_list_non_exhaustive()
+    {
+        true
+    } else {
+        false
+    }
+}
+
+/// Check if a block contains one of these:
+/// - Empty block with an expr (e.g., `{ Self::default() }`)
+/// - One statement (e.g., `{ return Self::default(); }`)
+fn check_block_calls_default(cx: &LateContext<'_>, block: &hir::Block<'_>) -> Option<Span> {
+    if let Some(expr) = block.expr
+        && block.stmts.is_empty()
+        && check_expr_call_default(cx, expr)
+    {
+        // Block only has a trailing expression, e.g. `Self::default()`
+        return None;
+    } else if let [hir::Stmt { kind, .. }] = block.stmts
+        && let hir::StmtKind::Expr(expr) | hir::StmtKind::Semi(expr) = kind
+        && let hir::ExprKind::Ret(Some(ret_expr)) = expr.kind
+        && check_expr_call_default(cx, ret_expr)
+    {
+        // Block has a single statement, e.g. `return Self::default();`
+        return None;
+    }
+
+    // trim first and last character, and trim spaces
+    let mut span = block.span;
+    span = span.with_lo(span.lo() + BytePos::from_usize(1));
+    span = span.with_hi(span.hi() - BytePos::from_usize(1));
+    span = trim_span(cx.sess().source_map(), span);
+
+    Some(span)
+}
+
+/// Check for `Self::default()` call syntax or equivalent
+fn check_expr_call_default(cx: &LateContext<'_>, expr: &hir::Expr<'_>) -> bool {
+    if let hir::ExprKind::Call(callee, &[]) = expr.kind
+        // FIXME: does this include `Self { }` style calls, which is equivalent,
+        //        but not the same as `Self::default()`?
+        // FIXME: what should the whole_call_expr (3rd arg) be?
+        && is_default_equivalent_call(cx, callee, None)
```

Comment:
From what I get from the code, if you pass `Some(expr)`, it should look inside `callee` and look if its content is similar to the default implementation.

In your case, since you only lint on `#[derive(Default)]`, I'm not sure it would make a difference, but it probably wouldn't hurt.